File: 0940-distinct-subsequences-ii/0940-distinct-subsequences-ii.py
class Solution:
    def distinctSubseqII(self, s: str) -> int:
        MOD = 10**9 + 7
        end_with = [0] * 26
        
        for char in s:
            idx = ord(char) - ord('a')
            # 1 (for sequence consisting of just `char`) + total existing distinct subsequences
            end_with[idx] = (sum(end_with) + 1) % MOD
            
        return sum(end_with) % MOD

File: 0940-distinct-subsequences-ii/0940-distinct-subsequences-ii.cpp
class Solution {
public:
    int distinctSubseqII(string s) {
        int MOD = 1e9 + 7;
        vector<long long> ends_with(26, 0);
        long long total_sum = 0;

        for (char c : s) {
            int idx = c - 'a';
            // New distinct subsequences ending in character 'c'
            long long new_count = (1 + total_sum) % MOD;
            
            // Update total sum by adding new count and subtracting old count for character 'c'
            total_sum = (total_sum + new_count - ends_with[idx] + MOD) % MOD;
            ends_with[idx] = new_count;
        }

        return total_sum;
    }
};